### ISSUE TITLE
[flang] Compilation error of explicit declaration of dummy argument

### ISSUE THREAD

**Issue by ohno-fj** [llvm/llvm-project#60224]
```
Version of flang-new : 16.0.0(77f2f34d696b77fe5bf05afbe7386966b6bcc8ba)
```

flang-new points out that there is no explicit type declaration of dummy argument.

The following are the test program, flang-new, gfortran and ifort compilation result.

snggm004_2.f90:
```fortran
PROGRAM main
  IMPLICIT NONE
  EXTERNAL    sub1
  INTEGER  :: i1 = 1
  integer  :: j
  CALL sub (sub1, i1)
  print *,'OK'
END PROGRAM main

SUBROUTINE sub (routine, i)
  IMPLICIT NONE
  INTEGER, INTENT(IN   ) :: i
  CALL routine (i)
END SUBROUTINE sub

SUBROUTINE sub1 (ii)
  IMPLICIT NONE
  INTEGER, INTENT(IN   ) :: ii
  WRITE (6,*)  ii
END SUBROUTINE sub1
```

```
$ flang-new -flang-experimental-exec snggm004_2.f90
error: Semantic errors in snggm004_2.f90
./snggm004_2.f90:10:17: error: No explicit type declared for 'routine'
  SUBROUTINE sub (routine, i)
                  ^^^^^^^
$
```

```
$ gfortran snggm004_2.f90
$
```

```
$ ifort snggm004_2.f90
$
```

**Comment by klausler** [comment]
Bug remains in current sources, and would be a good first project for somebody in the "community" to fix.

**Comment by shourya5** [comment]
I'd like to take this

**Comment by s-pratik** [comment]
Hii, I want to contribute can anyone help me how can I get started?

**Comment by omjavaid** [comment]
Hi s-pratik and shourya5 kindly assign this bug to you if you are interested in fixing it.

**Comment by s-pratik** [comment]
> Hi s-pratik and shourya5 kindly assign this bug to you if you are interested in fixing it.

Ya sure I am interested but I need some help to understand the issue properly

**Comment by luporl** [comment]
It seems the issue is that `routine`'s type not being explicitly declared should not be an error in this case, because of the `call` statement using it.
However, I'm having a hard time to find this in F2018 spec. If someone could point out where this is described it would help.

**Comment by kiranchandramohan** [comment]
In the standard (https://j3-fortran.org/doc/year/18/18-007r1.pdf), there is a section on Dummy procedure.
```
15.2.2.3 Dummy procedures
32 1 A dummy argument that is specified to be a procedure or appears as the procedure designator in a procedure
33 reference is a dummy procedure. A dummy procedure with the POINTER attribute is a dummy procedure
34 pointer.
```
There is some discussion about this in https://groups.google.com/g/comp.lang.fortran/c/X8SEt0Mx_X8/m/OeK7F0kJAgAJ.

**Comment by sscalpone** [comment]
Try with and without the IMPLICIT NONE in subroutine sub.  That might give a clue to why the error is raised.

**Comment by SahilPatidar** [comment]
is this issue still open? if yes and someone is not working on that then I'd like to take this one.

**Comment by luporl** [comment]
Yes, it's still open and so far no one is working on it. Please go ahead.

**Comment by SahilPatidar** [comment]
glad you replied, thanks for the info.

**Comment by SahilPatidar** [comment]
make me correct if I am wrong.
> It seems the issue is that `routine`'s type not being explicitly declared should not be an error in this case, because of the `call` statement using it.

which means that any dummy procedure doesn't need to be explicitly declared if a dummy procedure appears immediately following the word CALL in a CALL statement.

**Comment by luporl** [comment]
> which means that any dummy procedure doesn't need to be explicitly declared if a dummy procedure appears immediately following the word CALL in a CALL statement.

Yes, this is true, but see kiranchandramohan's comment and discussion link. There are also other cases in which dummy procedures don't need to be declared.

This part of the discussion sums up the rationale:
> If it is not declared as an array, then all that's left is a dummy function. Dummy subroutines are also recognized by their use in a call statement.

Initially, I thought that IMPLICIT NONE should make non declared subroutines an error, but that's not the case.
It just turns implicit types into errors. But as subroutines don't return a value, they still don't need to be declared.
An undeclared dummy function triggers an error because its (return) type is not explicitly declared.

You don't necessarily need to handle all possible cases to fix this particular issue, but it's nice to have the full picture in mind.

**Comment by klausler** [comment]
This discussion makes me think that this bug is probably not a very good "first issue".

The root cause here is that there's a symbol that appears only as a dummy argument and as an actual argument to a procedure with an implicit interface.  We don't know whether it is a procedure or an object, and actually don't have to care much -- during execution it will just be an address that has to be forwarded.  It could be represented today in f18's symbol table as a symbol that has EntityDetails.

There are two problems with that, however, which make this harder to solve than one might think.  First, EntityDetails symbols are used only in name resolution, and they are placeholders for symbols that will eventually become ObjectEntityDetails or ProcEntityDetails symbols once their usage makes clear what they are.  Any EntityDetails symbols that survive as such to the end of name resolution for a subprogram end up getting converted into objects more or less by default,, and that would have to change.

The second problem is that none of the rest of semantics or lowering ever expects to encounter a symbol table entry that is EntityDetails rather than one of its two subclasses.  If name resolution were to be changed so that EntityDetails symbols remained as such, fixing up the consequences would require detecting and upgrading all of the code sites downstream in compilation that would care.

One possible solution would be to change name resolution to default the EntityDetails symbols -- those that are only being forwarded from dummy arguments to actual arguments of procedures with implicit interfaces -- into ProcEntityDetails symbols, not ObjectEntityDetails symbols.  That might work, but it would be hard to figure out what's going on if it didn't work.

So I think this bug probably has to wait until a specialist can get in there and architect the best solution.

**Comment by klausler** [comment]
https://reviews.llvm.org/D150814
